### PR TITLE
Track size of row in RowContainer

### PR DESCRIPTION
**Pull request by oerling**
Adds a uint32_t to the row for tracking width of out of line variable
length data. This is only added if there are variable lenght fields or
accumulators.

This allows controlling the output batch size for hash join or group
by and for getting predictable size chunks of data for spilling.

Follow-up PRs will add setting this in hash build, group by key insert
and when growing variable width accumulators.

### CODE REVIEW

File: velox/exec/Aggregate.h [hunk 1]
```diff
@@ -56,10 +62,15 @@ class Aggregate {
   // @param offset Offset in bytes from the start of the row of the accumulator
   // @param nullByte Offset in bytes from the start of the row of the null flag
   // @param nullMask The specific bit in the nullByte that stores the null flag
-  void setOffsets(int32_t offset, int32_t nullByte, uint8_t nullMask) {
+  void setOffsets(
+      int32_t offset,
+      int32_t nullByte,
+      uint8_t nullMask,
+      int32_t rowSizeOffset) {
```
Comment (mbasmanova) [inline velox/exec/Aggregate.h:69]: It would be nice to add @param comment to document this parameter.

File: velox/exec/Aggregate.h [hunk 2]
```diff
@@ -139,6 +150,13 @@ class Aggregate {
   // 'groups'. No-op for fixed length accumulators.
   virtual void destroy(folly::Range<char**> /*groups*/) {}
 
+  // Returns the type that extractAccumulator will produce. This is
+  // different from resultType_ for complex final aggregations. Needed
+  // for e.g. spilling.
+  virtual TypePtr accumulatorType() const {
```
Comment (mbasmanova) [inline velox/exec/Aggregate.h:156]: It is confusing that the comment says "This is different from resultType_", but the implementation returns resultType_. It might be cleaner to not provide the default implementation for this method and require individual aggregations to specify this. Alternatively, Aggregate constructor could take accumulatorType as a parameter.

File: velox/exec/Aggregate.h [hunk 3]
```diff
@@ -190,6 +208,17 @@ class Aggregate {
     return numNulls_ && (group[nullByte_] & nullMask_);
   }
 
+  void incrementRowSize(char* row, uint64_t bytes) {
+    VELOX_DCHECK(rowSizeOffset_);
+    uint32_t* ptr = reinterpret_cast<uint32_t*>(row + rowSizeOffset_);
+    uint64_t size = *ptr + bytes;
+    *ptr = std::min<uint64_t>(bytes, std::numeric_limits<uint32_t>::max());
```
Comment (mbasmanova) [inline velox/exec/Aggregate.h:215]: My reading of this method is that bytes can be larger than 32-bit integer. Why is it ok to truncate it? Also, shouldn't we update ptr to "size", not "bytes". Are there tests for this code path?

File: velox/exec/Aggregate.h [hunk 4]
```diff
@@ -190,6 +208,17 @@ class Aggregate {
...
+  }
+
+  void setNull(char* group) {
```
Comment (mbasmanova) [inline velox/exec/Aggregate.h:218]: This method appears unused. What is it for? Can it be added together with first usage?

File: velox/exec/Aggregate.h [hunk 5]
```diff
@@ -241,6 +270,11 @@ class Aggregate {
   // Offset of fixed length accumulator state in group row.
   int32_t offset_;
 
+  // Offset of uint32_t row byte size of row. 0 if accumulator is
+  // fixed width. The size is capped at 4G and will stay at 4G and not
```
Comment (mbasmanova) [inline velox/exec/Aggregate.h:274]: Can you explain the motivation for 4G cap and why it works to truncate past that?

File: velox/exec/GroupingSet.cpp [hunk 6]
```diff
@@ -180,16 +180,19 @@ void GroupingSet::initializeGlobalAggregation() {
 
   // Row layout is:
   //  - null flags - one bit per aggregate,
+  // int32_t row size,
```
Comment (mbasmanova) [inline velox/exec/GroupingSet.cpp:183]: typo: missing "-"

File: velox/exec/GroupingSet.cpp [hunk 7]
```diff
@@ -180,16 +180,19 @@ void GroupingSet::initializeGlobalAggregation() {
...
   //  - fixed-width accumulators - one per aggregate
-  int32_t offset = bits::nbytes(aggregates_.size());
+  int32_t rowSizeOffset = bits::nbytes(aggregates_.size());
+  int32_t offset = rowSizeOffset + sizeof(int32_t);
```
Comment (mbasmanova) [inline velox/exec/GroupingSet.cpp:186]: The PR description says "This is only added if there are variable lenght fields or accumulators.", but this code suggests that rowSize field is always present.

Also, typo in PR description: lenght -> length

File: velox/exec/RowContainer.cpp [hunk 8]
```diff
@@ -152,6 +169,9 @@ RowContainer::RowContainer(
       }
     }
   }
+  if (hasProbedFlag) {
```
Comment (mbasmanova) [inline velox/exec/RowContainer.cpp:172]: Unrelated fix? Perhaps, mention it in the PR description.

File: velox/exec/RowContainer.h [hunk 9]
```diff
@@ -127,10 +131,23 @@ class RowContainer {
   // Allocates a new row and initializes possible aggregates to null.
   char* newRow();
 
+  uint32_t rowSize(const char* row) const {
+    return fixedRowSize_ +
+        (rowSizeOffset_
+             ? *reinterpret_cast<const uint32_t*>(row + rowSizeOffset_)
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:137]: reinterpret_cast<const uint32_t*>(row + rowSizeOffset_) is long and repeats multiple times. It would be nice to put it in a helper function.

File: velox/exec/RowContainer.h [hunk 10]
```diff
@@ -127,10 +131,23 @@ class RowContainer {
...
+             : 0);
+  }
+
   // Adds 'rows' to the free rows list and frees any associated
   // variable length data.
   void eraseRows(folly::Range<char**> rows);
 
+  void incrementRowSize(char* row, uint64_t bytes) {
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:145]: This logic is duplicated between Aggregate and RowContainer. Is there any particular reason group by doesn't use RowContainer?

File: velox/exec/RowContainer.h [hunk 11]
```diff
@@ -182,10 +199,20 @@ class RowContainer {
...
   }
 
-  // Extracts up to 'maxRows' rows starting at the position of 'iter'. A default
-  // constructed or reset iter starts at the beginning. Returns the number of
-  // rows written to 'rows'. Returns 0 when at end.
-  int32_t listRows(RowContainerIterator* iter, int32_t maxRows, char** rows);
+  // Extracts up to 'maxRows' rows starting at the position of
+  // 'iter'. A default constructed or reset iter starts at the
+  // beginning. Returns the number of rows written to 'rows'. Returns
+  // 0 when at end. Stops after te total size of returned rows exceeds
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:205]: typo: te -> the

File: velox/exec/RowContainer.h [hunk 12]
```diff
@@ -255,9 +288,28 @@ class RowContainer {
     return rows_.allocatedBytes() + stringAllocator_.retainedSize();
   }
 
+  // Returns the number of fixed size rows that can be allocated
+  // without growing the container and the number of unused bytes of
+  // reserved storage for variable length data.
+  std::pair<uint64_t, uint64_t> freeSpace() {
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:294]: std::pair<uint64_t, uint64_t> is not descriptive and it will be confusing to tell what the values are. Consider, adding a special-purpose struct with descriptive field names.

File: velox/exec/RowContainer.h [hunk 13]
```diff
@@ -298,6 +350,12 @@ class RowContainer {
   // Offset of the pointer to the next free row on a free row.
   static constexpr int32_t kNextFreeOffset = 0;
 
+  struct SpillRun {
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:353]: Does this belong in this PR? Would you document this struct?

File: velox/exec/RowContainer.h [hunk 14]
```diff
@@ -609,8 +667,12 @@ class RowContainer {
   std::vector<RowColumn> rowColumns_;
   // Bit position of probed flag, 0 if none.
   int32_t probedFlagOffset_ = 0;
+
   // Bit position of free bit.
-  int32_t freeFlagOffset_{0};
+  int32_t freeFlagOffset_ = 0;
+  int32_t nextFreeOffset_ = 0;
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:673]: nextFreeOffset_ appears unused. Remove.

File: velox/exec/RowContainer.h (hunk 10 above)
Comment (oerling) [reply velox/exec/RowContainer.h:145]: There are aggregates that are not in a RowContainer for global agg but might as well be in one.
More importantly, the aggregates do not knowabout RowContainer, onlly about HashStringAllocator. So duplicating this is more encapsulation than making RowContainer and Aggregtes know about each other.

**Comment by mbasmanova** [comment]
@oerling There are build failures:

```
In file included from ../../velox/exec/HashStringAllocator.cpp:17:
../.././velox/exec/RowContainer.h: In member function 'uint32_t& facebook::velox::exec::RowContainer::variableRowSize(char*)':
../.././velox/exec/RowContainer.h:388:47: error: invalid use of member function 'int32_t facebook::velox::exec::RowContainer::rowSizeOffset() const' (did you forget the '()' ?)
  388 |     return *reinterpret_cast<uint32_t*>(row + rowSizeOffset);
      |                                               ^~~~~~~~~~~~~
      |                                                            ()
```

**Review by mbasmanova — APPROVED** [review]
Thank you, Orri.

**Comment by oerling** [comment]
Can you accept the diff  <https://www.internalfb.com/diff/D31268602> D31268602 ?
